Source/DafnyCore/Verifier/BoogieGenerator.ExpressionWellformed.cs

```diff
@@ -781,16 +781,14 @@ private record ReadsCheckDelayer(ExpressionTranslator etran, Function selfCallsA
                     List<Expression> contextDecreases = codeContext.Decreases.Expressions;
                     List<Expression> calleeDecreases = e.Function.Decreases.Expressions;
                     if (e.Function == wfOptions.SelfCallsAllowance) {
-                      allowance = Bpl.Expr.True;
+                      allowance = Expression.CreateBoolLiteral(e.tok, true);
                       if (!e.Function.IsStatic) {
-                        allowance = BplAnd(allowance, Bpl.Expr.Eq(etran.TrExpr(e.Receiver), new Bpl.IdentifierExpr(e.tok, etran.This)));
+                        allowance = Expression.CreateAnd(allowance, Expression.CreateEq(e.Receiver, new ThisExpr(e.Function), e.Receiver.Type));
```

Comment:
It feels soo good to see Dafny expressions here. It will surely be helpful for making assertions explicit, but also possibly to consider another proving back-end.

Reply:
Yeah, I entirely agree. I sort of hope we can get to the point where every assertion in Boogie is a translation of an assertion in Dafny (which is perhaps synthetically constructed along the way). But that'll be a lot of refactoring, likely with significant impact on the shape of verification conditions and therefore brittleness.